File: src/main/java/com/kralofsky/citybikes/bot/util/ExternalAbility.java
package com.kralofsky.citybikes.bot.util;

import lombok.Builder;
import lombok.Singular;
import org.telegram.abilitybots.api.bot.BaseAbilityBot;
import org.telegram.abilitybots.api.objects.*;
import org.telegram.abilitybots.api.sender.DefaultSender;
import org.telegram.abilitybots.api.sender.MessageSender;
import org.telegram.abilitybots.api.sender.SilentSender;
import org.telegram.telegrambots.meta.api.objects.Update;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.function.Predicate;

public abstract class ExternalAbility {
    protected MessageSender sender;
    protected SilentSender silent;
    protected BaseAbilityBot bot;

    protected String DEFAULT;

    abstract protected AbilityOptions getOptions();

    abstract protected void action(MessageContext ctx);

    protected void post(MessageContext ctx){}

    protected List<Reply> replies(){
        return Collections.emptyList();
    }

    public void init(BaseAbilityBot bot, String defaultCommand) {
        this.bot = bot;
        this.sender = new DefaultSender(bot);
        this.silent = new SilentSender(sender);
        this.DEFAULT = defaultCommand;
    }

    public Ability getAbility(){
        AbilityOptions o = getOptions();

        final Ability.AbilityBuilder ab = Ability.builder()
                .name(o.name)
                .info(o.info)
                .locality(o.locality)
                .privacy(o.privacy)
                .input(o.input)
                .action(this::action)
                .post(this::post)
                .flag(o.flags.toArray(new Predicate[0]));

        replies().forEach(r -> ab.reply(r.action, r.conditions.toArray(new Predicate[0])));

        return ab.build();
    }

    @Builder
    public static class AbilityOptions{
        private String name;
        private String info;
        private Privacy privacy;
        private Locality locality;
        private int input;

        @Singular private Collection<Predicate<Update>> flags;
    }
}

File: src/main/java/com/kralofsky/citybikes/bot/util/ExternalAbility.kt
package com.kralofsky.citybikes.bot.util

import com.kralofsky.citybikes.bot.CitybikeTelegramBot
import org.telegram.abilitybots.api.bot.BaseAbilityBot
import org.telegram.abilitybots.api.objects.*
import org.telegram.abilitybots.api.sender.DefaultSender
import org.telegram.abilitybots.api.sender.MessageSender
import org.telegram.abilitybots.api.sender.SilentSender
import org.telegram.telegrambots.meta.api.objects.Update
import java.util.function.Predicate

abstract class ExternalAbility {
    protected var sender: MessageSender? = null
    protected var silent: SilentSender? = null
    protected var bot: BaseAbilityBot? = null

    protected abstract val options: AbilityOptions

    val ability: Ability
        get() {
            val o = options

            val ab = Ability.builder()
                    .name(o.name)
                    .info(o.info)
                    .locality(o.locality)
                    .privacy(o.privacy)
                    .input(o.input)
                    .action { this.action(it) }
                    .post { this.post(it) }
                    .flag(*o.flags.toTypedArray())

            replies().forEach { r -> ab.reply(r.action, *r.conditions.toTypedArray()) }

            return ab.build()
        }

    protected abstract fun action(ctx: MessageContext)

    protected fun post(ctx: MessageContext) {}

    protected open fun replies(): List<Reply> {
        return emptyList()
    }

    fun init(bot: BaseAbilityBot) {
        this.bot = bot
        this.sender = DefaultSender(bot)
        this.silent = SilentSender(sender)
    }


    data class AbilityOptions(
            var name: String = CitybikeTelegramBot.DEFAULT_COMMAND,
            var info: String = "",
            var privacy: Privacy = Privacy.PUBLIC,
            var locality: Locality = Locality.ALL,
            var input: Int = 0,
            var flags: MutableCollection<Predicate<Update>> = mutableListOf()
    ){
        fun name(name: String) = apply { this.name = name }
        fun info(info: String) = apply { this.info = info }
        fun privacy(privacy: Privacy) = apply { this.privacy = privacy }
        fun locality(locality: Locality) = apply { this.locality = locality }
        fun input(input: Int) = apply { this.input = input }
        fun flag(flag: Predicate<Update>) = apply { this.flags.add(flag) }
    }
}
